arangod/Pregel/Algos/AIR/GraphFormat.cpp

```diff
@@ -35,37 +40,160 @@ GraphFormat::GraphFormat(application_features::ApplicationServer& server,
+    if (key.isString()) {
+      VPackBuilder tmp;
+      VPackBuilder innerBuilder;
+      innerBuilder.openObject();
+      innerBuilder.add(key.copyString(), document.get(key.stringRef()));
+      innerBuilder.close();
+
+      if (!finalBuilder.slice().isObject()) {
+        finalBuilder.openObject();
+        finalBuilder.close();
+      }
+      VPackCollection::merge(tmp, finalBuilder.slice(), innerBuilder.slice(), true, false);
+      finalBuilder.clear();
+      finalBuilder.add(tmp.slice());
+    } else if (key.isArray()) {
+      std::vector<VPackStringRef> path;
+      size_t pathLength = key.length();
+      size_t iterationStep = 0;
+
+      VPackBuilder innerArrayBuilder;
+      innerArrayBuilder.openObject();  // open outer object
+
+      for (auto&& innerKey : VPackArrayIterator(key)) {
+        // TODO: optimize in the future - right now complicated to build a fresh object via key-path strings
+
+        if (!innerKey.isString()) {
+          TRI_ASSERT(false);  // TODO: Add type checking in deserializer
+        }
+        // build up path - will change in every iteration step
+        path.emplace_back(innerKey.stringRef());
+
+        if (iterationStep < (pathLength - 1)) {
+          innerArrayBuilder.add(innerKey);  //, VPackValue(VPackValueType::Object));
+          innerArrayBuilder.openObject();
+        } else {
+          innerArrayBuilder.add(innerKey.copyString(), document.get(path));
+        }
+
+        // get slice of each document depth
+        iterationStep++;
+      }
+
+      // now close all inner opened objects
+      for (size_t step = 0; step < (pathLength - 1); step++) {
+        innerArrayBuilder.close();
+      }
+
+      innerArrayBuilder.close();  // close outer object
+      VPackBuilder tmp;
+      VPackCollection::merge(tmp, finalBuilder.slice(),
+                             innerArrayBuilder.slice(), true, false);
+      finalBuilder.clear();
+      finalBuilder.add(tmp.slice());
+    } else {
+      TRI_ASSERT(false);
+    }
+  }
+}
+
 // Extract vertex data from vertex document into target
 void GraphFormat::copyVertexData(std::string const& documentId,
                                  arangodb::velocypack::Slice vertexDocument,
                                  vertex_type& targetPtr) {
-  targetPtr.reset(_vertexAccumulatorDeclarations, _customDefinitions,  documentId, vertexDocument, _vertexIdRange++);
+  if (_dataAccess.readVertex->slice().isArray()) {
+    // copy only specified keys/key-paths to document
+    VPackBuilder tmpBuilder;
+    filterDocumentData(tmpBuilder, _dataAccess.readVertex->slice(), vertexDocument);
+    targetPtr.reset(_vertexAccumulatorDeclarations, _customDefinitions,
+                    _dataAccess, documentId, tmpBuilder.slice(), _vertexIdRange++);
+  } else {
+    // copy all
+    targetPtr.reset(_vertexAccumulatorDeclarations, _customDefinitions,
+                    _dataAccess, documentId, vertexDocument, _vertexIdRange++);
+  }
```

Comment:
note that we have full control over the type of `vertex_data`, so we could get away with not having yet another builder here.